### ISSUE TITLE
[Question] Single Env vs SubprocVecEnv

### ISSUE THREAD

**Issue by danielstankw** [DLR-RM/stable-baselines3#660]
**Important Note: We do not do technical support, nor consulting** and don't answer personal questions per email.
Please post your question on the [RL Discord](https://discord.com/invite/xhfNqQv), [Reddit](https://www.reddit.com/r/reinforcementlearning/) or [Stack Overflow](https://stackoverflow.com/) in that case.

### Question
Hi,
I am running my single environment learning using PPO with default hyperparameters. 
The `total_timesteps` for learning is set to `40_000` steps. If I want to use for example 8 envs simultaneously for collecting the experience from the env, in order to make the learning process equivalent to running a single env for 40k steps, would I set learning to 40k steps as well (as in single env) or to `40_000/8=5000 steps` ?

Assuming same seed for the weight initialisation and for the env in case of single and multiple envs, I should obtain equivalent results?

In the hyperparameters the `n_epoch=10`, from my experience in the supervised learning the n_epoch is usually much bigger ex. 200-1000, whats the reason for choosing such a small number?

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
Hmm I thought these questions were covered by #632 ? 

`total_timesteps` counts total number of steps over all environments so you should keep it as `total_timesteps=40000` (but note that your results will probably be very different from running single env 40k steps vs. eight envs for 5k steps. PPO training is only robust with multiple parallel environments, usually.)

> In the hyperparameters the n_epoch=10, from my experience in the supervised learning the n_epoch is usually much bigger ex. 200-1000, whats the reason for choosing such a small number?

`n_epochs` refers to number of times all of the samples from rollouts are used to update the policy, before collecting new rollouts and repeating the procedure. More is not necessirely better as the training becomes more off-policy on each update, which usually is not good for training on-policy-like algorithms like PPO (see e.g. OpenAI Spinning Up on this).

**Comment by danielstankw** [comment]
I am asking again cause I did get different results when using multiple envs vs single env even though i have set same number of `timesteps` and `seeds`.
I still dont understand why are the results different.
What the reasoning behind different results?

**Comment by Miffyli** [comment]
With a single env you will have highly correlated samples in your rollout, since they are (likely) from one single episode, and is likely a bad representative of the dynamics of the full environment (since it is only a single episode). Doing updates on this data will bias the network to that one direction represented by that data.

With multiple envs you have a chance of having data from different episodes and different areas of the environment -> better representative of the full environment -> network updates are less biased, and generally, the training is more stable.

**Comment by danielstankw** [comment]
Great, thank you for explanation :)

**Comment by aaakhan** [comment]
Hi, I'd like to ask a relevant question regarding this:

> With multiple envs you have a chance of having data from different episodes and different areas of the environment -> better representative of the full environment -> network updates are less biased, and generally, the training is more stable.

As I understand it, this makes sense. However I have tried this on PPO in a custom env in which the typical episodes is about 12 steps, while n_steps is 150. I have found that using subprocenv with 3 environments leads to tremendous improvement both in performance (each results is the avg and std of 7 runs).

![parallels_2figs](https://user-images.githubusercontent.com/39007202/146067427-1796c7c1-9700-4f4f-a7cb-d3b85857b834.png)


I kept trying to understand how this makes such a difference, since each update will contain multiple full trajectories anyway (12 <<150). 

For the 3 env I changed n_steps from 150 to 160, but I don't expect this to be the main reason.

So should I scratch this as a lucky outcome, or is there something inherent in sb3 or gym that makes the random seeding of each environment affect sampling the action distribution somehow?

**Comment by Miffyli** [comment]
@aaakhan 

Note that with 3 environments ("processors", "processes") you will collect 150 x 3 steps in total instead of just 150. So there will be more updates per rollout (more updates to the network), which alone could explain the faster learning.

**Comment by aaakhan** [comment]
Thank you for the response.

Would this be true even if I keep batch_size and n_epochs the same when working on n_steps=150?

**Comment by Miffyli** [comment]
> Would this be true even if I keep batch_size and n_epochs the same when working on n_steps=150?

If you mean that the comparison with `n_steps = 150 / num_envs` would be true even if you keep `batch_size` and `n_epochs` the same, then yes :).

**Comment by aaakhan** [comment]
I see, so if I'm understanding this correctly then the results I got were because the agent had 150x3 steps, which led to more batches per rollout and therefore more updates. 

Thank you very much I thought there was something inherently off in my environment or in gym (much less likely).
